File: src/google/docs.ts
import { google, docs_v1 } from 'googleapis';
import { getOAuth2Client } from './auth';
import { logger } from '../logger';

function getDocs() {
  return google.docs({ version: 'v1', auth: getOAuth2Client() });
}

interface ImageSlot {
  driveFileId: string;
  publicUrl: string;
}

export async function fillDoc(
  docId: string,
  title: string,
  captionBody: string,
  hashtags: string[],
  images: ImageSlot[],
): Promise<void> {
  const docs = getDocs();

  // Phase A: text replacements
  await docs.documents.batchUpdate({
    documentId: docId,
    requestBody: {
      requests: [
        {
          replaceAllText: {
            containsText: { text: '{{TITLE}}', matchCase: true },
            replaceText: title,
          },
        },
        {
          replaceAllText: {
            containsText: { text: '{{CAPTION}}', matchCase: true },
            replaceText: captionBody,
          },
        },
        {
          replaceAllText: {
            containsText: { text: '{{HASHTAGS}}', matchCase: true },
            replaceText: hashtags.join(' '),
          },
        },
      ],
    },
  });

  // Phase B: image slots — process 10 down to 1 sequentially with re-fetch each time
  // to avoid index invalidation issues
  for (let n = 10; n >= 1; n--) {
    const placeholder = `{{IMAGE_${n}}}`;
    const slot = images[n - 1]; // undefined when n > imageCount

    const docContent = await docs.documents.get({ documentId: docId });
    const body = docContent.data.body!;

    if (slot) {
      // Replace the placeholder text with an inline image
      const range = findTextRange(body, placeholder);
      if (!range) {
        logger.warn(`Placeholder ${placeholder} not found in doc — skipping`);
        continue;
      }

      await docs.documents.batchUpdate({
        documentId: docId,
        requestBody: {
          requests: [
            {
              deleteContentRange: {
                range: { startIndex: range.start, endIndex: range.end },
              },
            },
            {
              insertInlineImage: {
                location: { index: range.start },
                uri: slot.publicUrl,
                // Let Docs pick dimensions; it will use the image's natural size capped by page width
                objectSize: undefined,
              },
            },
          ],
        },
      });
    } else {
      // Delete the entire block containing this placeholder
      const blockRange = findBlockRange(body, placeholder);
      if (!blockRange) {
        logger.warn(`Block for ${placeholder} not found — skipping`);
        continue;
      }

      await docs.documents.batchUpdate({
        documentId: docId,
        requestBody: {
          requests: [
            {
              deleteContentRange: {
                range: { startIndex: blockRange.start, endIndex: blockRange.end },
              },
            },
          ],
        },
      });
    }
  }
}

// Finds the start/end index of a literal text string inside the doc body.
function findTextRange(
  body: docs_v1.Schema$Body,
  text: string,
): { start: number; end: number } | null {
  for (const element of body.content ?? []) {
    const result = findInParagraph(element, text);
    if (result) return result;

    // Check inside table cells
    if (element.table) {
      for (const row of element.table.tableRows ?? []) {
        for (const cell of row.tableCells ?? []) {
          for (const cellElement of cell.content ?? []) {
            const r = findInParagraph(cellElement, text);
            if (r) return r;
          }
        }
      }
    }
  }
  return null;
}

function findInParagraph(
  element: docs_v1.Schema$StructuralElement,
  text: string,
): { start: number; end: number } | null {
  if (!element.paragraph) return null;
  for (const run of element.paragraph.elements ?? []) {
    const content = run.textRun?.content ?? '';
    const idx = content.indexOf(text);
    if (idx !== -1 && run.startIndex != null) {
      return {
        start: run.startIndex + idx,
        end: run.startIndex + idx + text.length,
      };
    }
  }
  return null;
}

// Finds the range of the smallest deletable block containing the placeholder.
// Strategy: find the paragraph (or table row) that contains the placeholder and
// return its full range including the trailing newline character.
function findBlockRange(
  body: docs_v1.Schema$Body,
  text: string,
): { start: number; end: number } | null {
  const content = body.content ?? [];

  for (const element of content) {
    // Paragraph-level block
    if (element.paragraph) {
      const found = findInParagraph(element, text);
      if (found && element.startIndex != null && element.endIndex != null) {
        return { start: element.startIndex, end: element.endIndex };
      }
    }

    // Table-level block: delete the entire row containing the placeholder
    if (element.table) {
      for (const row of element.table.tableRows ?? []) {
        let rowContainsPlaceholder = false;
        for (const cell of row.tableCells ?? []) {
          for (const cellElement of cell.content ?? []) {
            if (findInParagraph(cellElement, text)) {
              rowContainsPlaceholder = true;
            }
          }
        }
        if (rowContainsPlaceholder && row.startIndex != null && row.endIndex != null) {
          return { start: row.startIndex, end: row.endIndex };
        }
      }
    }
  }

  return null;
}

File: go/internal/google/docs.go
package google

import (
	"context"
	"fmt"
	"log/slog"
	"net/http"
	"strings"

	googledocs "google.golang.org/api/docs/v1"
	"google.golang.org/api/option"
)

func newDocsService(ctx context.Context, client *http.Client) (*googledocs.Service, error) {
	return googledocs.NewService(ctx, option.WithHTTPClient(client))
}

// FillDoc fills a Google Doc with the campaign content.
//
// Phase A: text replacements (one batchUpdate, four replaceAllText).
// Phase B: image slots 10→1, each with a fresh documents.get to avoid index
//          invalidation — identical to the TS implementation's explicit design choice.
func FillDoc(
	ctx context.Context,
	client *http.Client,
	docID, title, captionBody string,
	hashtags []string,
	images []ImageSlot,
) error {
	svc, err := newDocsService(ctx, client)
	if err != nil {
		return fmt.Errorf("docs service: %w", err)
	}

	// ── Phase A: text replacements ────────────────────────────────────────────
	_, err = svc.Documents.BatchUpdate(docID, &googledocs.BatchUpdateDocumentRequest{
		Requests: []*googledocs.Request{
			replaceText("{{TITLE}}", title),
			replaceText("{{CAPTION}}", captionBody),
			replaceText("{{HASHTAGS}}", strings.Join(hashtags, " ")),
			replaceText("{{NUMBER_OF_POSTS}}", fmt.Sprintf("%d", len(images))),
		},
	}).Context(ctx).Do()
	if err != nil {
		return fmt.Errorf("phase A batchUpdate: %w", err)
	}

	// ── Phase B: image slots 10 → 1 ──────────────────────────────────────────
	for n := 10; n >= 1; n-- {
		placeholder := fmt.Sprintf("{{IMAGE_%d}}", n)
		var slot *ImageSlot
		if n <= len(images) {
			s := images[n-1]
			slot = &s
		}

		// Re-fetch doc on every iteration to get fresh character indices.
		doc, err := svc.Documents.Get(docID).Context(ctx).Do()
		if err != nil {
			return fmt.Errorf("phase B get doc (slot %d): %w", n, err)
		}

		if slot != nil {
			// Replace placeholder text with inline image.
			r := findTextRange(doc.Body, placeholder)
			if r == nil {
				slog.Warn("placeholder not found in doc — skipping", "placeholder", placeholder)
				continue
			}
			_, err = svc.Documents.BatchUpdate(docID, &googledocs.BatchUpdateDocumentRequest{
				Requests: []*googledocs.Request{
					{
						DeleteContentRange: &googledocs.DeleteContentRangeRequest{
							Range: &googledocs.Range{
								StartIndex: r.start,
								EndIndex:   r.end,
							},
						},
					},
					{
						InsertInlineImage: &googledocs.InsertInlineImageRequest{
							Location: &googledocs.Location{Index: r.start},
							Uri:      slot.PublicURL,
						},
					},
				},
			}).Context(ctx).Do()
			if err != nil {
				return fmt.Errorf("phase B insert image (slot %d): %w", n, err)
			}
		} else {
			// Delete the entire block containing this placeholder.
			r := findBlockRange(doc.Body, placeholder)
			if r == nil {
				slog.Warn("block for placeholder not found — skipping", "placeholder", placeholder)
				continue
			}
			_, err = svc.Documents.BatchUpdate(docID, &googledocs.BatchUpdateDocumentRequest{
				Requests: []*googledocs.Request{
					{
						DeleteContentRange: &googledocs.DeleteContentRangeRequest{
							Range: &googledocs.Range{
								StartIndex: r.start,
								EndIndex:   r.end,
							},
						},
					},
				},
			}).Context(ctx).Do()
			if err != nil {
				return fmt.Errorf("phase B delete block (slot %d): %w", n, err)
			}
		}
	}

	return nil
}

// ── Helpers ──────────────────────────────────────────────────────────────────

func replaceText(find, replace string) *googledocs.Request {
	return &googledocs.Request{
		ReplaceAllText: &googledocs.ReplaceAllTextRequest{
			ContainsText: &googledocs.SubstringMatchCriteria{
				Text:      find,
				MatchCase: true,
			},
			ReplaceText: replace,
		},
	}
}

type charRange struct {
	start int64
	end   int64
}

// findTextRange locates the exact character range of a literal string inside the doc body.
// Searches paragraphs at top level and inside table cells — mirrors findTextRange in docs.ts.
func findTextRange(body *googledocs.Body, text string) *charRange {
	for _, elem := range body.Content {
		if r := findInParagraph(elem, text); r != nil {
			return r
		}
		if elem.Table != nil {
			for _, row := range elem.Table.TableRows {
				for _, cell := range row.TableCells {
					for _, cellElem := range cell.Content {
						if r := findInParagraph(cellElem, text); r != nil {
							return r
						}
					}
				}
			}
		}
	}
	return nil
}

func findInParagraph(elem *googledocs.StructuralElement, text string) *charRange {
	if elem.Paragraph == nil {
		return nil
	}
	for _, run := range elem.Paragraph.Elements {
		if run.TextRun == nil {
			continue
		}
		content := run.TextRun.Content
		idx := strings.Index(content, text)
		if idx != -1 {
			return &charRange{
				start: run.StartIndex + int64(idx),
				end:   run.StartIndex + int64(idx) + int64(len(text)),
			}
		}
	}
	return nil
}

// findBlockRange returns the range of the smallest deletable block (paragraph
// or table row) that contains the placeholder text.
//
// IMPORTANT: endIndex is decremented by 1 to exclude the trailing newline —
// the Docs API rejects deleteContentRange when the range includes the trailing
// newline of a paragraph or table row. This matches the TS findBlockRange exactly.
func findBlockRange(body *googledocs.Body, text string) *charRange {
	for _, elem := range body.Content {
		if elem.Paragraph != nil {
			if findInParagraph(elem, text) != nil {
				return &charRange{
					start: elem.StartIndex,
					end:   elem.EndIndex - 1,
				}
			}
		}
		if elem.Table != nil {
			for _, row := range elem.Table.TableRows {
				rowContains := false
				for _, cell := range row.TableCells {
					for _, cellElem := range cell.Content {
						if findInParagraph(cellElem, text) != nil {
							rowContains = true
						}
					}
				}
				if rowContains {
					return &charRange{
						start: row.StartIndex,
						end:   row.EndIndex - 1,
					}
				}
			}
		}
	}
	return nil
}
